src/pwm_input.rs

```diff
@@ -0,0 +1,373 @@
+    /// This will limit the frequency range where the timer can operate.
+    ///
+    /// The minimum frequency that can be sampled is 90% the provided frequency
+    DutyCycle(T),
+
+    /// In this mode an algorithm calculate the optimal value for the autoreload register and the
+    /// prescaler register in order to be able to sample signal with a frequency higher than the
+    /// provided value : there is no margin.
+    RawFrequency(T),
+
+    /// In this mode, the user gives the desired arr and presc value that are then directly
+    /// programmed in the register.
+    RawValues { arr: u16, presc: u16 },
+}
+
+pub trait PwmInputExt: Sized {
+    fn pwm_input<PINS, T>(
+        self,
+        PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> PwmInput<Self, PINS>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>;
+}
+
+impl PwmInputExt for TIM2 {
+    fn pwm_input<PINS, T>(
+        self,
+        pins: PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> PwmInput<Self, PINS>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>,
+    {
+        mapr.mapr()
+            .modify(|_, w| unsafe { w.tim2_remap().bits(PINS::REMAP) });
+        tim2(self, pins, clocks, apb, mode, dbg)
+    }
+}
+
+impl PwmInputExt for TIM3 {
+    fn pwm_input<PINS, T>(
+        self,
+        pins: PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> PwmInput<Self, PINS>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>,
+    {
+        mapr.mapr()
+            .modify(|_, w| unsafe { w.tim3_remap().bits(PINS::REMAP) });
+        tim3(self, pins, clocks, apb, mode, dbg)
+    }
+}
+
+impl PwmInputExt for TIM4 {
+    fn pwm_input<PINS, T>(
+        self,
+        pins: PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> PwmInput<Self, PINS>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>,
+    {
+        mapr.mapr()
+            .modify(|_, w| w.tim4_remap().bit(PINS::REMAP == 1));
+        tim4(self, pins, clocks, apb, mode, dbg)
+    }
+}
+
+
+fn get_score(x:u32, y:u32) -> u32 {
+   if x > y {
+      x - y
+   } else {
+      y - x
+   }
+}
+/// F_timer = PCLK / [ (ARR+1) * (PRESC+1) ]
+/// We want maximum resolution, so ARR = u16::MAX;
+/// <=> PRESC = PCLK / [ u16::MAX * F_timer ] -1
+fn compute_arr_presc(freq: u32, clock: u32) -> (u16, u16) {
+    let mut arr: u32 = (core::u16::MAX - 1) as u32;
+    let mut presc: u32;
+
+    let mut curr_best_presc : u32 = 0;
+    let mut curr_best_arr : u32 = 0;
+    let mut curr_best_distance : u32= core::u32::MAX;
+
+    for _ in 0..65530 {
```

Comment:
We are looking for a better algorithm, but this algorithm at least works.
Maybe we should mention in the documentation the fact that it is expensive to let the code find the best arr and presc value.

Before merging this PR we should also run a test were we try every combination possible.

Reply:
Could you bisect (binary search) this instead? That'd only be 16 iterations at worst I think?

Reply:
Yes, this is what we plan to implement !

Reply:
Makes sense, if it is linear, binary search does sound like a good idea, otherwise a slow algorithm with a comment mentioning its slowness sounds good enough for now imo

Reply:
What's the typical inputs for this function? I can search a better method.

Reply:
Something between 1 and 12 000 000 for the frequency, and 8 000 000 up to 72 000 000 for the clock.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=515abfa4cd97d0274f66562bb0477851

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8220a82864f38d668481f69ec49b3d1b

TLDR: can't fail:

```rust
fn compute_arr_presc(freq: u32, clock: u32) -> (u16, u16) {
    if freq == 0 {
        return (core::u16::MAX, core::u16::MAX);
    }
    let presc = clock / freq.saturating_mul(core::u16::MAX as u32 + 1);
    let mut arr = clock / freq.saturating_mul(presc + 1);
    if arr > 0 {
        arr -= 1;
    }
    (arr as u16, presc as u16)
}
```

Reply:
If I understand correctly the RM 15.3.2 p.369

> n upcounting mode, the counter counts from 0 to the auto-reload value (content of the TIMx_ARR register), then restarts from 0 and generates a counter overflow event.

The formula is f = clock / (arr * (presc + 1)) (no `-1` to arr)

Thus the decrement in the code above is useless.

It's also said that arr = 0 will block the timer to not evoluting. That's what my function returns is freq > clock. Maybe returning `max(1, arr)` would be better.

So, this might be better:

```rust
fn compute_arr_presc(freq: u32, clock: u32) -> (u16, u16) {
    if freq == 0 {
        return (core::u16::MAX, core::u16::MAX);
    }
    let presc = clock / freq.saturating_mul(core::u16::MAX as u32 + 1);
    let arr = clock / freq.saturating_mul(presc + 1);
    (core::cmp::max(1, arr as u16), presc as u16)
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6cd00739ab7af5a763c598d52a3b3208

Reply:
Thank you @TeXitoi, I replaced our algorithm by yours !